app/lib/exchange-rate/money.ts

```diff
@@ -0,0 +1,480 @@
+        },
+      },
+    ],
+  },
+};
+
+type MoneyData = {
+  /**
+   * Currency of the money
+   */
+  currency: string;
+  /**
+   * Amount of the money stored in the `unit` format
+   */
+  amount: Big;
+  /**
+   * Unit in which the `amount` is stored. It is always the minimal supported unit for the `currency` (e.g. millisatoshi
+   * for BTC)
+   */
+  amountUnit: CurrencyUnit;
+  /**
+   * Unit of the initial amount that was provided when creating the money
+   */
+  initialUnit: CurrencyUnit;
+};
+
+export class Money {
+  private readonly _data: MoneyData;
+
+  constructor(data: MoneyInput) {
+    const { baseUnit, minUnit, selectedUnit } =
+      Money.getCurrencyDataForInput(data);
+    const unit = selectedUnit ?? baseUnit;
+
+    const multiplier = unit.factor.div(minUnit.factor);
+    const amountInMinUnit = new Big(data.amount).mul(multiplier);
+
+    this._data = {
+      amount: amountInMinUnit.round(minUnit.decimals, Big.roundDown),
+      currency: data.currency,
+      amountUnit: minUnit,
+      initialUnit: unit,
+    };
+
+    Object.freeze(this);
+    Object.freeze(this._data);
+  }
+
+  static create(props: MoneyInput): Money {
+    return new Money(props);
+  }
+
+  /**
+   * Sum an array of moneys.
+   *
+   * If the array is empty, a currency must be specified so that 0 can be returned in that currency.
+   *
+   * The precision, etc., is based on the first item in the array.
+   * If the array is empty, the options object will be used instead.
+   */
+  static sum(moneys: Money[], currency?: string): Money {
+    if (moneys.length === 0 && currency === undefined) {
+      throw new Error(
+        "Currency must be set when summing an empty list of money's",
+      );
+    }
+
+    if (moneys.length === 0 && currency !== undefined) {
+      return new Money({ amount: 0, currency });
+    }
+
+    return moneys.slice(1).reduce((sum, money) => sum.add(money), moneys[0]);
+  }
+
+  static max(moneys: Money[]): Money {
+    if (moneys.length === 0) {
+      throw new Error('Need at least one money for comparison');
+    }
+    return moneys.reduce(
+      (max, money) => (money.greaterThan(max) ? money : max),
+      moneys[0],
+    );
+  }
+
+  static min(moneys: Money[]): Money {
+    if (moneys.length === 0) {
+      throw new Error('Need at least one money for comparison');
+    }
+    return moneys.reduce(
+      (min, money) => (money.lessThan(min) ? money : min),
+      moneys[0],
+    );
+  }
+
+  /**
+   * Compare two money objects.
+   *
+   * 1 if money1 is greater than money2
+   * 0 if equal
+   * -1 if money1 is less than money2
+   *
+   * This can be plugged directly into array.sort(),
+   * and it will cause the array to be sorted in ascending order.
+   */
+  static compare(money1: Money, money2: Money): number {
+    money1.assertSameCurrency(money2);
+    return money1.amount().cmp(money2.amount());
+  }
+
+  get currency(): string {
+    return this._data.currency;
+  }
+
```

Comment:
I think we should also have a getter for the symbol too. In most cases we will just use toLocaleString, but for showing the balance I want to be able to style the symbol separate from the string in some JSX

Reply:
it can't be a getter because money object represents all possible units for the currency and units can have different symbols. but we can make it function like this:
```
/**
   * Returns the currency symbol for the requested unit
   * @param unit Unit of currency. Default value is base unit (e.g. btc or usd)
   */
  getCurrencySymbol(unit?: string): string {
    return this.getCurrencyUnit(unit).symbol;
  }
 ```

Reply:
added this